src/lib.rs

```diff
@@ -0,0 +1,389 @@
+        Ok(FMatcher {
+            ptn: self.ptn,
+            options: self.options,
+        })
+    }
+}
+
+/// The fuzzy matcher.
+pub struct FMatcher<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMatcher<'a> {
+    /// A convenience method that automatically builds a pattern for you using `FMBuilder`'s
+    /// default options.
+    pub fn new(ptn: &'a str) -> Result<FMatcher, Box<dyn Error>> {
+        FMBuilder::new(ptn)?.build()
+    }
+
+    /// Does this fuzzy matcher match `text`? Returns `Ok(())` on success or `Err(usize)` on error
+    /// where the `usize` is the line number of the first line in `text` where the match failed.
+    pub fn matches(&self, text: &str) -> Result<(), usize> {
+        let mut names = HashMap::new();
+        let mut ptn_lines = self.ptn.lines();
+        let mut ptnl = ptn_lines.next();
+        let mut text_lines = text.lines();
+        let mut textl = text_lines.next();
+        let mut text_lines_off = 1;
+        loop {
+            match (ptnl, textl) {
+                (Some(x), Some(y)) => {
+                    if x.trim() == WILDCARD {
+                        ptnl = ptn_lines.next();
+                        match ptnl {
+                            Some(x) => {
+                                while let Some(y) = textl {
+                                    if self.match_line(&mut names, x, y) {
+                                        break;
+                                    }
+                                    textl = text_lines.next();
+                                    text_lines_off += 1;
+                                }
+                                text_lines_off -= 1;
+                            }
+                            None => return Ok(()),
+                        }
+                    } else if self.match_line(&mut names, x, y) {
+                        ptnl = ptn_lines.next();
+                        textl = text_lines.next();
+                        text_lines_off += 1;
+                    } else {
+                        return Err(text_lines_off);
+                    }
+                }
+                (None, None) => return Ok(()),
+                (Some(x), None) if x.trim() == WILDCARD => return Ok(()),
+                (Some(_), None) => return Err(text_lines_off),
+                (None, Some(_)) => return Err(text_lines_off),
+            }
+        }
+    }
+
+    fn match_line<'b>(
+        &self,
+        names: &mut HashMap<&'a str, &'b str>,
+        mut ptn: &'a str,
+        mut text: &'b str,
+    ) -> bool {
+        if self.options.ignore_leading_whitespace {
+            ptn = ptn.trim_start();
+            text = text.trim_start();
+        } else {
+            todo!();
+        }
+
+        if self.options.ignore_trailing_whitespace {
+            ptn = ptn.trim_end();
+            text = text.trim_end();
+        } else {
+            todo!();
+        }
+
+        let sww = ptn.starts_with(WILDCARD);
```

Comment:
Have you been hanging around with physicists again? :P

Reply:
They're too clever for me!